In [2]:
from functools import cache
from itertools import combinations
import os

import pandas as pd
from tinydb import Query

import citutils.dice_utils as du
from citutils.database import DB


/var/folders/83/_0vtzb0d2c1bj8sxc04n21m80000gn/T/ipykernel_88518/363497145.py:5: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [3]:
db = DB(os.path.join("source.json"))

skill = Query().meta_tags.any("skill")
basic =  Query().meta_tags.any("basic")



print(db.create_query_text_section(query = basic & skill & Query().meta_tags.any("str"))) # not clear any of it should be 


#### Butchery I  
**Requirements:** 1 STR  
**Cost:** 5 SP  
When you roll matching dice with a [Severing] weapon, do 1 crit damage and apply apply an [injury] that disables one of the enemy's limbs.  

#### Butchery II  
**Requirements:** 3 STR, Butchery I  
**Cost:** 5 SP  
Replaces Butchery I. When you roll a half crit with a [Severing] weapon, do 1 crit damage and apply an [injury] that disables one of the enemy's limbs. On full crit, remove the limb.  

#### Charge  
**Cost:** 5 SP  
Pay 1 AP. Move double your movement speed in a straight line. You must move at least two spaces. Once you reach an enemy, you may perform an attack, paying AP as usual. Before you roll your attack you can choose to add your STR to the damage or to your attack roll. If you are entering a STR contest of some kind, add your STR to your roll.  

#### Chokehold  
**Requirements:** {Grappling Adept}  
**Cost:** 4 SP  
You must be grappling your target and your target must be [prone] for you to perform a Cho

In [8]:
print(df.to_markdown(index=False))


|   adv_level |   total_rolls |   passing_matching |   passing_halfcrit |   matching |   halfcrit |   fullcrit |    prob_pm |   prob_phc |   prob_m |    prob_hc |     prob_fc |
|------------:|--------------:|-------------------:|-------------------:|-----------:|-----------:|-----------:|-----------:|-----------:|---------:|-----------:|------------:|
|           3 |          7776 |               2912 |               4571 |       6230 |       4651 |       1526 | 0.374486   | 0.587834   | 0.801183 | 0.598122   | 0.196245    |
|           2 |          1296 |                336 |                639 |        795 |        671 |        171 | 0.259259   | 0.493056   | 0.613426 | 0.517747   | 0.131944    |
|           1 |           216 |                 32 |                 79 |         80 |         91 |         16 | 0.148148   | 0.365741   | 0.37037  | 0.421296   | 0.0740741   |
|           0 |            36 |                  2 |                  7 |          5 |         11 |          1 | 0.

|   adv_level |   total_rolls |   passing_matching |   passing_halfcrit |   matching |   halfcrit |   fullcrit |    prob_pm |   prob_phc |   prob_m |    prob_hc |     prob_fc |
|------------:|--------------:|-------------------:|-------------------:|-----------:|-----------:|-----------:|-----------:|-----------:|---------:|-----------:|------------:|
|           3 |          7776 |               2912 |               4571 |       6230 |       4651 |       1526 | 0.374486   | 0.587834   | 0.801183 | 0.598122   | 0.196245    |
|           2 |          1296 |                336 |                639 |        795 |        671 |        171 | 0.259259   | 0.493056   | 0.613426 | 0.517747   | 0.131944    |
|           1 |           216 |                 32 |                 79 |         80 |         91 |         16 | 0.148148   | 0.365741   | 0.37037  | 0.421296   | 0.0740741   |
|           0 |            36 |                  2 |                  7 |          5 |         11 |          1 | 0.0555556  | 0.194444   | 0.138889 | 0.305556   | 0.0277778   |
|          -1 |           216 |                  5 |                 10 |         35 |         16 |          1 | 0.0231481  | 0.0462963  | 0.162037 | 0.0740741  | 0.00462963  |
|          -2 |          1296 |                 12 |                 13 |        225 |         21 |          1 | 0.00925926 | 0.0100309  | 0.173611 | 0.0162037  | 0.000771605 |
|          -3 |          7776 |                 27 |                 16 |       1355 |         26 |          1 | 0.00347222 | 0.00205761 | 0.174254 | 0.00334362 | 0.000128601 |

In [ ]:
import json

def add_meta_tag(data, obj_name, meta_tag):
    if isinstance(data, dict):
        if obj_name in data:
            # Add meta_tag to the current object
            data[obj_name]["meta_tags"].append(meta_tag)
        
        for key, value in data.items():
            # Recursively go through nested objects
            add_meta_tag(value, obj_name, meta_tag)
    elif isinstance(data, list):
        # Recursively go through each item in the list
        for item in data:
            add_meta_tag(item, obj_name, meta_tag)

# Your JSON data
json_data = {
    "new": {
        "items": {
            "the_braggart": {
                "name": "The Braggart",
                "effect": "A cloak with bright silver stitching.  When worn, DC 9 illusion of a full set of plate armor.",
                "meta_tags": ["item"]
            },
            # ... (other items)
        },
        # ... (other sections)
    }
}

# Example usage: Add "weapon_skill" meta_tag to all meta_tags inside "weapon_skills"
add_meta_tag(json_data, "weapon_skills", "weapon_skill")

# Pretty print the updated JSON
print(json.dumps(json_data, indent=2))